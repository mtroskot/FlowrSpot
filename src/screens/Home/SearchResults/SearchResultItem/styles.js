import { StyleSheet } from 'react-native';
import { dimensions, fonts } from 'src/styles';
const { rem } = dimensions;
const imageSize = 50 * rem;

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    backgroundColor: '#fff',
    padding: 5 * rem
  },
  avatarView: {
    flex: 0.2,
    paddingLeft: 20 * rem
  },
  image: {
    height: imageSize,
    width: imageSize,
    borderRadius: imageSize / 2,
    resizeMode: 'cover'
  },
  contentView: {
    justifyContent: 'center',
    flex: 0.8
  },
  mainContentText: {
    ...fonts.style.normal,
    fontSize: 16 * rem,
    color: '#000'
  },
  contentText: {
    ...fonts.style.italic,
    fontSize: 12 * rem,
    marginTop: 5 * rem,
    color: '#666666'
  }
});

export default styles;
