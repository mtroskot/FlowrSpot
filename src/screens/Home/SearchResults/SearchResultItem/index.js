import React from 'react';
import { Image, Text, TouchableOpacity, View } from 'react-native';
import PropTypes from 'prop-types';
import styles from 'src/screens/Home/SearchResults/SearchResultItem/styles';

const SearchResultItem = props => {
  const { id, name, latin_name, profile_picture, onResultPress } = props;
  return (
    <TouchableOpacity activeOpacity={0.95} onPress={() => onResultPress(id)} style={styles.container}>
      <View style={styles.avatarView}>
        <Image source={{ uri: `https:${profile_picture}` }} style={styles.image} />
      </View>
      <View style={styles.contentView}>
        <Text style={styles.mainContentText}>{name}</Text>
        <Text style={styles.contentText}>{latin_name}</Text>
      </View>
    </TouchableOpacity>
  );
};

SearchResultItem.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  latin_name: PropTypes.string.isRequired,
  profile_picture: PropTypes.string.isRequired,
  onResultPress: PropTypes.func.isRequired
};

export default React.memo(SearchResultItem);
