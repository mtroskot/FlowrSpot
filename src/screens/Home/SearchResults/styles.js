import { StyleSheet } from 'react-native';
import { dimensions, fonts } from 'src/styles';
const { rem } = dimensions;

const styles = StyleSheet.create({
  flatListContentContainer: {
    borderRadius: 10 * rem,
    overflow: 'hidden'
  },
  separator: {
    borderColor: '#cecece',
    borderWidth: 1,
    height: 1
  },
  triangle: {
    marginLeft: 25 * rem,
    width: 0,
    height: 0,
    backgroundColor: 'transparent',
    borderStyle: 'solid',
    borderLeftWidth: 8 * rem,
    borderRightWidth: 8 * rem,
    borderBottomWidth: 16 * rem,
    borderLeftColor: 'transparent',
    borderRightColor: 'transparent',
    borderBottomColor: 'white'
  },
  messageView: {
    backgroundColor: 'white',
    padding: 20 * rem,
    borderRadius: 10 * rem
  },
  loadingView: {
    padding: 10 * rem
  },
  messageText: {
    ...fonts.style.normal,
    color: 'black',
    textAlign: 'center',
    fontSize: 14 * rem
  },
  closeButton: {
    width: '100%',
    position: 'absolute',
    backgroundColor: '#e8e8e8',
    padding: 10 * rem,
    borderRadius: 10 * rem
  },
  closeText: {
    ...fonts.style.normal,
    textAlign: 'center',
    fontSize: 14 * rem
  }
});

export default styles;
