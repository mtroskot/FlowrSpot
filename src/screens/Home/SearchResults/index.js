import React from 'react';
import { FlatList, Text, View } from 'react-native';
import { CustomButton, ListItemSeparator, Loader } from 'src/components';
import SearchResultItem from 'src/screens/Home/SearchResults/SearchResultItem';
import PropTypes from 'prop-types';
import styles from 'src/screens/Home/SearchResults/styles';

const SearchResults = props => {
  const { searchData, isSearching, onResultPress, onClosePress } = props;
  const { searchResults, message } = searchData;

  if (message || isSearching) {
    return (
      <>
        <View style={styles.triangle} />
        <View style={styles.messageView}>
          {isSearching ? (
            <Loader size={'large'} viewStyle={styles.loadingView} />
          ) : (
            <Text style={styles.messageText}>{message}</Text>
          )}
        </View>
      </>
    );
  }
  return (
    <>
      <View style={styles.triangle} />
      <View>
        <FlatList
          data={searchResults}
          ItemSeparatorComponent={ListItemSeparator}
          renderItem={({ item }) => {
            const { id, name, latin_name, profile_picture } = item;
            return <SearchResultItem {...{ id, name, latin_name, profile_picture, onResultPress }} />;
          }}
          contentContainerStyle={styles.flatListContentContainer}
          bounces={false}
          keyExtractor={item => item.id.toString()}
        />
      </View>
      <View>
        <CustomButton
          onPress={onClosePress}
          tOpacityStyle={styles.closeButton}
          activeOpacity={0.95}
          text={'Close'}
          textStyle={styles.closeText}
        />
      </View>
    </>
  );
};

SearchResults.propTypes = {
  searchData: PropTypes.shape({
    searchResults: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
    message: PropTypes.string.isRequired
  }).isRequired,
  isSearching: PropTypes.bool.isRequired,
  onResultPress: PropTypes.func.isRequired,
  onClosePress: PropTypes.func.isRequired
};

export default React.memo(SearchResults);
