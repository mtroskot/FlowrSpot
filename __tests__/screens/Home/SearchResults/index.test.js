import React from 'react';
import { shallow } from 'enzyme';
import SearchResults from 'src/screens/Home/SearchResults';
import SearchResultItem from 'src/screens/Home/SearchResults/SearchResultItem';

describe('SearchResults wrapper', () => {
  it('renders loading indicator in search results view', () => {
    const props = {
      searchData: {
        searchResults: [],
        message: null
      },
      isSearching: true,
      onResultPress: jest.fn(),
      onClosePress: jest.fn()
    };

    const wrapper = shallow(<SearchResults {...props} />);
    expect(wrapper.first().type()).toEqual(React.Fragment);
    expect(wrapper.find('Fragment')).toHaveLength(1);
    expect(wrapper.find('View')).toHaveLength(2);
    expect(wrapper.find('Memo(Loader)')).toHaveLength(1);
    expect(wrapper.find('Text')).toHaveLength(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders message in search results view', () => {
    const props = {
      searchData: {
        searchResults: [],
        message: 'Something went wrong'
      },
      isSearching: false,
      onResultPress: jest.fn(),
      onClosePress: jest.fn()
    };

    const wrapper = shallow(<SearchResults {...props} />);
    expect(wrapper.first().type()).toEqual(React.Fragment);
    expect(wrapper.find('Fragment')).toHaveLength(1);
    expect(wrapper.find('View')).toHaveLength(2);
    expect(wrapper.find('Memo(Loader)')).toHaveLength(0);
    expect(wrapper.find('Text')).toHaveLength(1);
    expect(
      wrapper
        .find('Text')
        .first()
        .props().children
    ).toEqual(props.searchData.message);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders search results', () => {
    const props = {
      searchData: {
        searchResults: [
          { id: 'id1', name: 'name1', latin_name: 'latin_name1', profile_picture: 'profile_picture1' },
          { id: 'id2', name: 'name2', latin_name: 'latin_name2', profile_picture: 'profile_picture2' }
        ],
        message: null
      },
      isSearching: false,
      onResultPress: jest.fn(),
      onClosePress: jest.fn()
    };

    const wrapper = shallow(<SearchResults {...props} />);
    expect(wrapper.first().type()).toEqual(React.Fragment);
    expect(wrapper.find('Fragment')).toHaveLength(1);
    expect(wrapper.find('View')).toHaveLength(3);
    expect(wrapper.find('Memo(Loader)')).toHaveLength(0);
    expect(wrapper.find('Text')).toHaveLength(1);
    expect(
      wrapper
        .find('Text')
        .first()
        .props().children
    ).toEqual('Close');
    expect(wrapper.find('TouchableOpacity')).toHaveLength(1);
    expect(wrapper.find('FlatList')).toHaveLength(1);
    expect(wrapper.find('FlatList').props().data).toEqual(props.searchData.searchResults);
    expect(
      wrapper
        .find('FlatList')
        .props()
        .keyExtractor({ id: 'id' })
    ).toEqual('id');
    const item = props.searchData.searchResults[0];
    expect(
      wrapper
        .find('FlatList')
        .props()
        .renderItem({ item })
    ).toEqual(
      <SearchResultItem
        {...{
          id: item.id,
          name: item.name,
          latin_name: item.latin_name,
          profile_picture: item.profile_picture,
          onResultPress: props.onResultPress
        }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});

describe('interaction', () => {
  const props = {
    searchData: {
      searchResults: [
        { id: 'id1', name: 'name1', latin_name: 'latin_name1', profile_picture: 'profile_picture1' },
        { id: 'id2', name: 'name2', latin_name: 'latin_name2', profile_picture: 'profile_picture2' }
      ],
      message: null
    },
    isSearching: false,
    onResultPress: jest.fn(),
    onClosePress: jest.fn()
  };
  afterEach(() => {
    jest.clearAllMocks();
  });
  const wrapper = shallow(<SearchResults {...props} />);
  const RenderItem = wrapper.find('FlatList').prop('renderItem');
  const renderItemShallowWrapper = shallow(<RenderItem item={props.searchData.searchResults[0]} />);

  it('calling SearchResultItem onResultPress, should call the onResultPress callback', () => {
    renderItemShallowWrapper
      .find('Memo(SearchResultItem)')
      .first()
      .prop('onResultPress')();
    expect(props.onResultPress).toHaveBeenCalledTimes(1);
    expect(props.onClosePress).toHaveBeenCalledTimes(0);
  });
  it('pressing Close buttton, should call the onClosePress callback', () => {
    wrapper
      .find('TouchableOpacity')
      .first()
      .prop('onPress')();
    expect(props.onClosePress).toHaveBeenCalledTimes(1);
    expect(props.onResultPress).toHaveBeenCalledTimes(0);
  });
});
