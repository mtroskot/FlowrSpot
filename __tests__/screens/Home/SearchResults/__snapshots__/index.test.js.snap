// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults wrapper renders loading indicator in search results view 1`] = `
<Fragment>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "borderBottomColor": "white",
        "borderBottomWidth": 31.57894736842105,
        "borderLeftColor": "transparent",
        "borderLeftWidth": 15.789473684210526,
        "borderRightColor": "transparent",
        "borderRightWidth": 15.789473684210526,
        "borderStyle": "solid",
        "height": 0,
        "marginLeft": 49.34210526315789,
        "width": 0,
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 19.736842105263158,
        "padding": 39.473684210526315,
      }
    }
  >
    <Memo(Loader)
      size="large"
      viewStyle={
        Object {
          "padding": 19.736842105263158,
        }
      }
    />
  </View>
</Fragment>
`;

exports[`SearchResults wrapper renders message in search results view 1`] = `
<Fragment>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "borderBottomColor": "white",
        "borderBottomWidth": 31.57894736842105,
        "borderLeftColor": "transparent",
        "borderLeftWidth": 15.789473684210526,
        "borderRightColor": "transparent",
        "borderRightWidth": 15.789473684210526,
        "borderStyle": "solid",
        "height": 0,
        "marginLeft": 49.34210526315789,
        "width": 0,
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 19.736842105263158,
        "padding": 39.473684210526315,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
          "fontFamily": "Ubuntu-Regular",
          "fontSize": 27.63157894736842,
          "textAlign": "center",
        }
      }
    >
      Something went wrong
    </Text>
  </View>
</Fragment>
`;

exports[`SearchResults wrapper renders search results 1`] = `
<Fragment>
  <View
    style={
      Object {
        "backgroundColor": "transparent",
        "borderBottomColor": "white",
        "borderBottomWidth": 31.57894736842105,
        "borderLeftColor": "transparent",
        "borderLeftWidth": 15.789473684210526,
        "borderRightColor": "transparent",
        "borderRightWidth": 15.789473684210526,
        "borderStyle": "solid",
        "height": 0,
        "marginLeft": 49.34210526315789,
        "width": 0,
      }
    }
  />
  <View>
    <FlatList
      ItemSeparatorComponent={[Function]}
      bounces={false}
      contentContainerStyle={
        Object {
          "borderRadius": 19.736842105263158,
          "overflow": "hidden",
        }
      }
      data={
        Array [
          Object {
            "id": "id1",
            "latin_name": "latin_name1",
            "name": "name1",
            "profile_picture": "profile_picture1",
          },
          Object {
            "id": "id2",
            "latin_name": "latin_name2",
            "name": "name2",
            "profile_picture": "profile_picture2",
          },
        ]
      }
      disableVirtualization={false}
      horizontal={false}
      initialNumToRender={10}
      keyExtractor={[Function]}
      maxToRenderPerBatch={10}
      numColumns={1}
      onEndReachedThreshold={2}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      updateCellsBatchingPeriod={50}
      windowSize={21}
    />
  </View>
  <View>
    <TouchableOpacity
      activeOpacity={0.95}
      onPress={[MockFunction]}
      style={
        Object {
          "backgroundColor": "#e8e8e8",
          "borderRadius": 19.736842105263158,
          "padding": 19.736842105263158,
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "fontFamily": "Ubuntu-Regular",
            "fontSize": 27.63157894736842,
            "textAlign": "center",
          }
        }
      >
        Close
      </Text>
    </TouchableOpacity>
  </View>
</Fragment>
`;
